### ISSUE TITLE
XPCServices in the framework...

### ISSUE THREAD

**Issue by digitalmoksha** [zorgiepoo/sparkle-ui-xpc-issues#41]
I was wondering if it makes sense to, by default, put the xpc services into their standard location in the framework: `Sparkle.framework/Versions/A/XPCServices`, and with `Sparkle.framework/XCPService` linked to  `Versions/Current/XPCServices`.  From what I can tell, this seems to be how most system frameworks organize it.

And maybe `make release` builds two frameworks, `Sparkle.framework` and `Sparkle-sandboxed.framework` - then it's obvious which one to use and there is less wrangling needed for the developer's build.  I would think the downloader service would not be included by default, since I think most modern apps would have the `com.apple.security.network.client` entitlement.

The reason this came up for me is because I use RubyMotion to build my app, and it will automatically sign the XPC services inside a framework if they're located in the standard place.  And I would prefer to do the `make release` and simply copy the `Sparkle-sandboxed.framework` into place without wrangling the xpc services.

**Comment by digitalmoksha** [comment]
ok, I see the issue, based on #39. They really do need to be signed differently than the main app.

Can the main framework be signed with the sandbox entitlement? I'm still having trouble getting it to work - I get the dialog indicating a new version exists, but it fails when it goes to install the update.  I'll keep poking at it, must have signed something wrong...

**Comment by zorgiepoo** [comment]
I thought long and hard about this in the early development. Sparkle follows the same approach as Growl for making the XPC services optional and not apart of the framework; 3rd party frameworks can't get away doing the same thing as system frameworks here.

The tooling for "Code Sign on Copy" for XPC services simply isn't there for Xcode. Xcode really doesn't like having the services inside a framework that you copy in your app. Also as you figured out, the services need to be signed differently than the main application.

It's a bit of an assumption to say most apps have a network entitlement. In fact, one could argue that if sandboxed apps need entitlements, they should be separated into XPC services. The main application (and Sparkle) should be sandboxed but not have entitlements.

Fyi, I probably won't be able to respond/contribute in a few days from now. Best to report issues in the official Spark repo where this branch has been pulled.

**Comment by digitalmoksha** [comment]
That's why I figured I better jump on this and try to get it working in my app before you're unavailable :-)

**Comment by zorgiepoo** [comment]
> It's a bit of an assumption to say most apps have a network entitlement. In fact, one could argue that if sandboxed apps need entitlements, they should be separated into XPC services. The main application (and Sparkle) should be sandboxed but not have entitlements.

It's been a while since I've looked at this so I have misspoken.. what I meant to say is that if sandboxed apps need *exceptions*, they should be separated into XPC services. Anything that is sandboxed needs entitlements. Note this branch contains several non-sandboxed XPC services (i.e, no entitlements used when signing) with the exception of the downloader service.

**Comment by digitalmoksha** [comment]
Thank you for the clarification.

I verified that indeed the xpc services should exist in an `XPCServices` folder in `MyApp.app/Contents`. I tried placing them in the `Sparkle.framework` itself and signing them correctly. However verifying the signature of the app produces an error: `unsealed contents present in the root directory of an embedded framework. In subcomponent: ....Contents/Frameworks/Sparkle.framework`

With them in the right place, it works now.
